File: backend/src/lib/commentEvents.ts
import type { Response } from 'express';

/**
 * Targeted SSE delivery registry.
 *
 * Maps userId → Set<Response> so we can push events to specific users
 * (e.g. notify only the customer who owns a comment, or only staff/admin).
 *
 * One user can have multiple tabs open → Set of responses per userId.
 */
const clientMap = new Map<string, Set<Response>>();

export function registerSSEClient(userId: string, res: Response): void {
  if (!clientMap.has(userId)) clientMap.set(userId, new Set());
  clientMap.get(userId)!.add(res);
}

export function unregisterSSEClient(userId: string, res: Response): void {
  const set = clientMap.get(userId);
  if (!set) return;
  set.delete(res);
  if (set.size === 0) clientMap.delete(userId);
}

/** Send a JSON event payload to a single user (all their open tabs). */
export function sendSSEToUser(userId: string, event: object): void {
  const clients = clientMap.get(userId);
  if (!clients || clients.size === 0) return;
  const payload = `data: ${JSON.stringify(event)}\n\n`;
  for (const res of clients) {
    try {
      res.write(payload);
    } catch {
      // Connection already closed; will be cleaned up on 'close' event
    }
  }
}

/** Broadcast a JSON event payload to multiple users at once. */
export function sendSSEToUsers(userIds: string[], event: object): void {
  if (userIds.length === 0) return;
  const payload = `data: ${JSON.stringify(event)}\n\n`;
  for (const userId of userIds) {
    const clients = clientMap.get(userId);
    if (!clients) continue;
    for (const res of clients) {
      try {
        res.write(payload);
      } catch {
        // ignore
      }
    }
  }
}

// ─── Event type interfaces ──────────────────────────────────────────────────

export interface CommentCreatedEvent {
  eventType: 'comment.created';
  comment: {
    id: string;
    menuItemId: string;
    menuItemName: string;
    customerName: string | null;
    content: string;
    rating: number | null;
    createdAt: string;
  };
}

export interface CommentRepliedEvent {
  eventType: 'comment.replied';
  commentId: string;
  reply: {
    id: string;
    content: string;
    staffName: string | null;
    createdAt: string;
  };
}

export interface NotificationCreatedEvent {
  eventType: 'notification.created';
  notification: {
    id: string;
    type: string;
    title: string;
    body: string;
    isRead: boolean;
    refId: string | null;
    createdAt: string;
  };
}

export type CommentRealtimeEvent =
  | CommentCreatedEvent
  | CommentRepliedEvent
  | NotificationCreatedEvent;

File: backend/app/events/comment_events.py
import asyncio
import json
from typing import Any

class CommentEventBus:
    """SSE event bus for comment/notification events targeted to specific users."""

    def __init__(self):
        # Map user_id -> list of Queues
        self._clients: dict[str, list[asyncio.Queue]] = {}

    def subscribe(self, user_id: str) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue()
        if user_id not in self._clients:
            self._clients[user_id] = []
        self._clients[user_id].append(q)
        return q

    def unsubscribe(self, user_id: str, q: asyncio.Queue):
        if user_id in self._clients:
            try:
                self._clients[user_id].remove(q)
            except ValueError:
                pass
            if not self._clients[user_id]:
                del self._clients[user_id]

    async def publish_to_user(self, user_id: str, event: dict[str, Any]):
        if user_id in self._clients:
            for q in list(self._clients[user_id]):
                try:
                    await q.put(event)
                except Exception:
                    self.unsubscribe(user_id, q)


comment_event_bus = CommentEventBus()
